torch/csrc/api/include/torch/nn/modules/container/parameterdict.h

```diff
@@ -0,0 +1,157 @@
+#pragma once
+
+#include <torch/nn/cloneable.h>
+#include <torch/ordered_dict.h>
+
+#include <vector>
+
+namespace torch {
+namespace nn {
+
+class ParameterDictImpl : public Cloneable<ParameterDictImpl> {
+ public:
+  using Iterator = OrderedDict<std::string, Tensor>::Iterator;
+  using ConstIterator = OrderedDict<std::string, Tensor>::ConstIterator;
+
+  ParameterDictImpl() = default;
+
+  explicit ParameterDictImpl(
+      torch::OrderedDict<std::string, torch::Tensor> params) {
+    parameters_.reserve(params.size());
+    for (const auto& item : params) {
+      insert(std::move(item.key()), std::move(item.value()));
+    }
+  }
+
+  /// `reset()` is empty for `ParameterDict`, since it does not have parameters
+  /// of its own.
+  void reset() override {}
+
+  /// Pretty prints the `ParameterDict` module into the given `stream`.
+  void pretty_print(std::ostream& stream) const override {
+    stream << "torch::nn::ParameterDict(" << std::endl;
+    for (const auto& pair : parameters_) {
+      stream << "(" << pair.key() << ")"
+             << ": Parameter containing: [" << pair.value().type()
+             << " of size " << pair.value().sizes() << "]";
+      ;
+      stream << std::endl;
+    }
+    stream << ")";
+  }
+
+  /// Insert the parameter along with the key into ParameterDict
+  /// The parameter is set to be require grad by default
+  Tensor& insert(std::string key, Tensor param, bool requires_grad = true) {
+    return register_parameter(key, param, requires_grad);
+  }
+
+  /// Remove key from the ParameterDict, and return its value
+  Tensor pop(const std::string& key) {
+    TORCH_CHECK(
+        parameters_.contains(key),
+        "No Parameter with name `",
+        key,
+        "` is registered");
+    torch::Tensor v = parameters_[key].clone();
```

Comment:
You don't need to clone(), we just need to return the same tensor, if you clone() the underlying storage will be copied. See below, it is not copied.
https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/container.py#L557